pydantic/_internal/_std_types_schema.py

```diff
@@ -82,25 +82,32 @@ def to_enum(__input_value: Any, _: core_schema.ValidationInfo) -> Enum:
     )
     metadata = build_metadata_dict(js_metadata=js_metadata)
 
+    lax: CoreSchema
     if issubclass(enum_type, int):
         # this handles `IntEnum`, and also `Foobar(int, Enum)`
         js_metadata['extra_updates'] = {'type': 'integer'}
-        return core_schema.chain_schema(
-            core_schema.int_schema(), literal_schema, core_schema.function_plain_schema(to_enum), metadata=metadata
+        lax = core_schema.chain_schema(
+            core_schema.int_schema(), literal_schema, core_schema.function_plain_schema(to_enum)
         )
     elif issubclass(enum_type, str):
         # this handles `StrEnum` (3.11 only), and also `Foobar(str, Enum)`
         # TODO: add test for StrEnum in 3.11, and also for enums that inherit from str/int
         js_metadata['extra_updates'] = {'type': 'string'}
-        return core_schema.chain_schema(
-            core_schema.str_schema(), literal_schema, core_schema.function_plain_schema(to_enum), metadata=metadata
+        lax = core_schema.chain_schema(
+            core_schema.str_schema(), literal_schema, core_schema.function_plain_schema(to_enum)
         )
     else:
-        return core_schema.function_after_schema(literal_schema, to_enum, metadata=metadata)
+        lax = core_schema.function_after_schema(literal_schema, to_enum)
+
+    return core_schema.lax_or_strict_schema(
+        lax_schema=lax,
+        strict_schema=core_schema.is_instance_schema(enum_type),
```

Comment:
either here or in a separate issue we need to think about the validation error you'd get.

I guess ideally the error should be the same or consistent between lax and strict.

We could use something like

```suggestion
        strict_schema=core_schema.custom_error_schema(
            core_schema.is_instance_schema(enum_type),
            'literal_error',
            custom_error_context=', '.join(... enum members),
        ),
```

not a blocker for this, but if we don't change now, we should create an issue to remember about this.

Reply:
also just realised that this won't work for JSON. 

I guess we'll need to use a function after validator with

```py
core_schema.is_instance_schema(enum_type, json_types={'int', 'float', 'str'})
```

I guess ultimately we might need a `core_schema.python_or_json_schema` type so we can have a custom schema for json data, but I think that's not necessary now.

Reply:
Here's what I'm getting with a custom error:

```
{
    'type': 'decimal_parsing',
    'loc': (
        'a',
        'lax-or-strict[lax=function-after[DecimalValidator(allow_inf_nan=False, check_digits=False, strict=False)(), union[is-instance[Decimal],int,float,constrained-str]],strict=custom-error[function-after[DecimalValidator(allow_inf_nan=False, check_digits=False, strict=False)(), is-instance[Decimal]]]]',
    ),
    'msg': 'Input should be a valid decimal',
    'input': 'foobar',
}
```

In particular `lax-or-strict` should never show up in an error loc. Unlike a union, we are either lax or strict for the entire validation. So `lax-or-strict` should just surface up the lax or strict error depending on which mode is being used.